libs/template/schema_test.go

```diff
@@ -0,0 +1,274 @@
+	assert.ErrorContains(t, err, "expected type boolean, but value is 1")
+	err = validateType("abc", jsonschema.BooleanType)
+	assert.ErrorContains(t, err, "expected type boolean, but value is \"abc\"")
+
+	// assert validation failing for string
+	err = validateType(int(1), jsonschema.StringType)
+	assert.ErrorContains(t, err, "expected type string, but value is 1")
+	err = validateType(float64(1), jsonschema.StringType)
+	assert.ErrorContains(t, err, "expected type string, but value is 1")
+	err = validateType(false, jsonschema.StringType)
+	assert.ErrorContains(t, err, "expected type string, but value is false")
+}
+
+func TestTemplateSchemaValidateConfig(t *testing.T) {
+	// define schema for config
+	jsonSchema := testSchema(t)
+
+	// define the config
+	config := map[string]any{
+		"int_val":    1,
+		"float_val":  1.1,
+		"bool_val":   true,
+		"string_val": "abc",
+	}
+
+	err := validateConfigValueTypes(config, jsonSchema)
+	assert.NoError(t, err)
+}
+
+func TestTemplateSchemaValidateConfigFailsForUnknownField(t *testing.T) {
+	// define schema for config
+	jsonSchema := testSchema(t)
+
+	// define the config
+	config := map[string]any{
+		"foo":        1,
+		"float_val":  1.1,
+		"bool_val":   true,
+		"string_val": "abc",
+	}
+
+	err := validateConfigValueTypes(config, jsonSchema)
+	assert.ErrorContains(t, err, "foo is not defined as an input parameter for the template")
+}
+
+func TestTemplateSchemaValidateConfigFailsForWhenIncorrectTypes(t *testing.T) {
+	// define schema for config
+	jsonSchema := testSchema(t)
+
+	// define the config
+	config := map[string]any{
+		"int_val":    1,
+		"float_val":  1.1,
+		"bool_val":   "true",
+		"string_val": "abc",
+	}
+
+	err := validateConfigValueTypes(config, jsonSchema)
+	assert.ErrorContains(t, err, "incorrect type for bool_val. expected type boolean, but value is \"true\"")
+}
+
+func TestTemplateSchemaValidateConfigFailsForWhenMissingInputParams(t *testing.T) {
+	// define schema for config
+	schemaJson := `{
+		"properties": {
+			"int_val": {
+				"type": "integer"
+			},
+			"string_val": {
+				"type": "string"
+			}
+		}
+		}`
+	var jsonSchema jsonschema.Schema
+	err := json.Unmarshal([]byte(schemaJson), &jsonSchema)
+	require.NoError(t, err)
+
+	// define the config
+	config := map[string]any{
+		"int_val": 1,
+	}
+
+	err = assignDefaultConfigValues(config, &jsonSchema)
+	assert.ErrorContains(t, err, "input parameter string_val is not defined in config")
+}
+
+func TestTemplateDefaultAssignment(t *testing.T) {
+	// define schema for config
+	schemaJson := `{
+		"properties": {
+			"foo": {
+				"type": "integer",
+				"default": 1
+			}
+		}
+		}`
+	var jsonSchema jsonschema.Schema
+	err := json.Unmarshal([]byte(schemaJson), &jsonSchema)
+	require.NoError(t, err)
+
+	// define the config
+	config := map[string]any{}
+
+	err = assignDefaultConfigValues(config, &jsonSchema)
+	assert.NoError(t, err)
+	assert.Equal(t, 1.0, config["foo"])
```

Comment:
Shouldn't this be an integer?

Reply:
Right now to read the schema, we do the following steps in order:
1. assign default values
2. cast any floats to ints if required
3. validate the types for all values

This test just checks that the assignment was done correctly, the casting then happens in step 2. Not casting the value in the assignment function helps keep the logic simple and is redundant if we are calling the cast function later